Completion production/injection rates/totals; Field keywords

jokva opened the pull request:
A small set of the completion family of keywords, water/oil/gas
production and injection. The tests and example data file are updated
accordingly, with edge cases.

Requires https://github.com/Ensembles/ert/pull/1167 and https://github.com/OPM/opm-parser/pull/881 (bugfixes). 

note: I've added support for a set of Field keywords to this PR. Easily split up obviously, but they're along the same lines so I figured  it's fine.

jokva committed:
Completion production/injection rates/totals

A small set of the completion family of keywords, water/oil/gas
production and injection. The tests and example data file are updated
accordingly, with edge cases.

jokva committed:
Cuts/ratios only apply to production

In line with Eclipse water cuts, gas-oil ratio etc. only applies to
production values, rather than all rates. This sparked some other
refactoring that moves the negative-else-zero logic into the rate
calculation.

jokva committed:
[Summary] Rate and rate-derived F-keyword

Support for rate or rate-derived keywords (i.e. sans pressure, sales
etc.) from the Field family of keywords.

jokva committed:
Explanations in the Summary test deck

A few short lines to make the summary test deck more understandable.

joakim-hove commented on opm/output/eclipse/Summary.cpp:
```diff
@@ -77,13 +77,14 @@ struct quantity {
+ * the index of the block in question. wells is simulation data.
  */
 struct fn_args {
     const std::vector< const Well* >& ecl_wells;
```
As in [this well](https://github.com/OPM/opm-parser/blob/master/opm/parser/eclipse/EclipseState/Schedule/Well.hpp) - maybe `schedule_wells` would be a better name?

jokva replied:
Good one, I'll update in the morning.

joakim-hove commented on opm/output/eclipse/Summary.cpp:
```diff
@@ -102,18 +103,51 @@ measure rate_unit< rt::gas >() { return measure::gas_surface_rate; }
+        const auto& name = ecl_well->name();
+        if( args.wells.wells.count( name ) == 0 ) continue;
+        const auto v = args.wells.at( name ).rates.get( phase, 0.0 );
+        if( ( v > 0 ) == injection )
```
I would spend less mental energy on ` if ( ( v > 0) && injection)`.

joakim-hove replied:
There are four states here `{ ( v >  0) , ( v < 0 )} x { true , false }` - but I assume only `{(v > 0) , true }` and `{ (v < 0) , false }` are valid - otherwise you have a massive internal error? So I guess you just do `v += sum`?

jokva replied:
Pretty much. These are the cases:
1. Injection is requested and this is an injecting well (rates are positive).
2. Injection is requested and this is a producer.
3. Production is requested and this is a producer.
4. Production is requested and this is an injector.

Which gives us:
1. `v > 0 == true && injection == true` -> sum increases
2. `v > 0 == true && injection != true )` -> no-op
3. `v > 0 == false && injection != true` -> sum increases
4. `v > 0 == false && injection == true` no-op

So it's a no-op when the well type doesn't matche the boolean parameter.

joakim-hove commented on opm/output/eclipse/Summary.cpp:
```diff
@@ -102,18 +103,51 @@ measure rate_unit< rt::gas >() { return measure::gas_surface_rate; }
+
+template< rt phase > inline quantity prodrate( const fn_args& args )
+{ return rate< phase, false >( args ); }
+template< rt phase > inline quantity injerate( const fn_args& args )
```
A blank line maybe?

jokva replied:
Maybe. I really wish C++ had reasonable function aliasing. Figured it wasn't worth spending too much space on.

joakim-hove commented on opm/output/eclipse/Summary.cpp:
```diff
@@ -102,18 +103,51 @@ measure rate_unit< rt::gas >() { return measure::gas_surface_rate; }
 }
 
+template< rt phase, bool injection = true >
+inline quantity crate( const fn_args& args ) {
```
For the hierarchy `well -> group -> field` I guess it makes sense to express everything as sums over well-lists, but when it comes to the connections/completions I do not see the point. The `Cxxx` variables are always only interesting as variables for one completion - and are never used as elements in a sum constituting a "larger object" - the way wells are summed into groups. Does this make sense?

jokva replied:
Yeah. I mean, we can throw in an assert I suppose, but the big reason for just reusing everything was that wells already come in a list, so it would be zero or one iteration. Could probably use less noise though, I agree with that, so I'll look into it.

jokva replied:
I've updated this. It's actually slightly longer (I **really** wish we had monads), but it does communicate the no-sum property of the completion keywords.

joakim-hove commented:
I will force merge this: OPM/opm-simulators#763 tomorrow and restart Travis.

jokva commented:
Excellent. Works on my machine, but you never know. I'll address the issues tomorrow.

jokva committed:
Rename ecl_wells -> schedule_wells

jokva committed:
Completion rate function over unit, not list

There are no such thing as sum-over-completions or over multiple wells.
Treat them as such to communicate this, rather than sum-over-list.
